src/lib.rs

```diff
@@ -356,33 +370,97 @@ impl KaboodleInner {
             log::info!("RECV [{sender}] {msg:?}");
 
             match msg {
-                SwimMessage::Ack(peer) => {
-                    // Insert the peer into our known_peers map with None as their "suspected since"
-                    // timestamp; if they were already in there with a suspected since timestamp,
-                    // this will reset them back to being non-suspected.
+                SwimMessage::Ack(peer, their_fingerprint, their_num_peers) => {
+                    // Insert the peer into our known_peers map as Known; if they were already in
+                    // there in a WaitingFor... state, this will reset them back to being known.
                     let mut known_peers = self.known_peers.lock().await;
                     known_peers.insert(peer, PeerState::Known(Instant::now()));
+                    let our_fingerprint =
+                        generate_fingerprint(&known_peers)[0..SHORT_FINGERPRINT_LENGTH].to_owned();
+                    let our_num_peers = known_peers.len() as u32;
                     drop(known_peers);
 
                     if let Some(observers) = self.curious_peers.remove(&peer) {
                         // Some of our peers were waiting to hear back about this ping
                         for observer in observers {
                             // todo: run these in parallel
-                            self.send_msg(&observer, &SwimMessage::Ack(peer)).await;
+                            self.send_msg(
+                                &observer,
+                                &SwimMessage::Ack(peer, our_fingerprint.to_owned(), our_num_peers),
+                            )
+                            .await;
                         }
                     }
+
+                    self.maybe_sync_known_peers(
+                        peer,
+                        our_fingerprint,
+                        our_num_peers,
+                        their_fingerprint,
+                        their_num_peers,
+                    )
+                    .await;
                 }
                 SwimMessage::KnownPeers(peers) => {
                     let mut known_peers = self.known_peers.lock().await;
-                    let now = Instant::now();
+                    let timestamp = Instant::now().checked_sub(MAX_PEER_SHARE_AGE).unwrap();
```

Comment:
If you think it's silly to call checked_sub and then unwrap it immediately, I agree! Clippy made me replace `Instant::now() - MAX_PEER_SHARE_AGE` with this. Once I get error handling in place, this will look less goofy.